src/dbnode/storage/shard.go

```diff
@@ -2660,74 +2661,217 @@ func (s *dbShard) Repair(
 
 func (s *dbShard) AggregateTiles(
 	ctx context.Context,
-	reader fs.DataFileSetReader,
 	sourceNsID ident.ID,
-	sourceBlockStart time.Time,
-	sourceShard databaseShard,
+	sourceShardID uint32,
+	blockReaders []fs.DataFileSetReader,
+	sourceBlockVolumes []shardBlockVolume,
 	opts AggregateTilesOptions,
-	wOpts series.WriteOptions,
+	targetSchemaDescr namespace.SchemaDescr,
 ) (int64, error) {
-	latestSourceVolume, err := sourceShard.latestVolume(sourceBlockStart)
+	blockReadersToClose := make([]fs.DataFileSetReader, 0, len(blockReaders))
+	defer func() {
+		for _, reader := range blockReadersToClose {
+			reader.Close()
+		}
+	}()
+
+	maxEntries := 0
+	for sourceBlockPos, blockReader := range blockReaders {
+		sourceBlockVolume := sourceBlockVolumes[sourceBlockPos]
+		openOpts := fs.DataReaderOpenOptions{
+			Identifier: fs.FileSetFileIdentifier{
+				Namespace:   sourceNsID,
+				Shard:       sourceShardID,
+				BlockStart:  sourceBlockVolume.blockStart,
+				VolumeIndex: sourceBlockVolume.latestVolume,
+			},
+			FileSetType:      persist.FileSetFlushType,
+			StreamingEnabled: true,
+		}
+
+		if err := blockReader.Open(openOpts); err != nil {
+			s.logger.Error("blockReader.Open", zap.Error(err))
+			return 0, err
+		}
+		if blockReader.Entries() > maxEntries {
+			maxEntries = blockReader.Entries()
+		}
+
+		blockReadersToClose = append(blockReadersToClose, blockReader)
+	}
+
+	crossBlockReader, err := fs.NewCrossBlockReader(blockReaders, s.opts.InstrumentOptions())
 	if err != nil {
+		s.logger.Error("NewCrossBlockReader", zap.Error(err))
 		return 0, err
 	}
+	defer crossBlockReader.Close()
 
-	openOpts := fs.DataReaderOpenOptions{
-		Identifier: fs.FileSetFileIdentifier{
-			Namespace:   sourceNsID,
-			Shard:       sourceShard.ID(),
-			BlockStart:  sourceBlockStart,
-			VolumeIndex: latestSourceVolume,
+	tileOpts := tile.Options{
+		FrameSize:          xtime.UnixNano(opts.Step.Nanoseconds()),
```

Comment:
FrameSize should be `time.Duration` not `xtime.UnixNano` right? They are both uint64 but time.Duration is for a time width vs an actual epoch time (xtime.UnixNano).

Reply:
👍